accord-core/src/main/java/accord/local/Commands.java

```diff
@@ -610,247 +665,236 @@ private static boolean maybeExecute(SafeCommandStore safeStore, SafeCommand safe
+    public enum Truncate { NO, TRUNCATE, ERASE }
+
+    public static Command setTruncated(SafeCommandStore safeStore, SafeCommand safeCommand)
     {
-        Command.Committed command = dependencySafeCommand.current().asCommitted();
-        Command dependency = safeDependency.current();
-        Invariants.checkState(dependency.hasBeen(PreCommitted));
-        if (dependency.hasBeen(Invalidated))
-        {
-            logger.trace("{}: {} is invalidated. Stop listening and removing from waiting on commit set.", command.txnId(), dependency.txnId());
-            safeDependency.removeListener(command.asListener());
-            waitingOn.removeWaitingOnCommit(dependency.txnId());
-            return true;
-        }
-        else if (dependency.executeAt().compareTo(command.executeAt()) > 0)
-        {
-            // dependency cannot be a predecessor if it executes later
-            logger.trace("{}: {} executes after us. Stop listening and removing from waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.removeWaitingOn(dependency.txnId(), dependency.executeAt());
-            safeDependency.removeListener(command.asListener());
-            return true;
-        }
-        else if (dependency.hasBeen(Applied))
-        {
-            logger.trace("{}: {} has been applied. Stop listening and removing from waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.removeWaitingOn(dependency.txnId(), dependency.executeAt());
-            safeDependency.removeListener(command.asListener());
-            return true;
-        }
-        else if (command.isWaitingOnDependency())
-        {
-            logger.trace("{}: adding {} to waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.addWaitingOnApply(dependency.txnId(), dependency.executeAt());
-            waitingOn.removeWaitingOnCommit(dependency.txnId());
-            return false;
-        }
-        else
+        return setTruncated(safeStore, safeCommand, Truncate.TRUNCATE, true);
+    }
+
+    public static Command setTruncated(SafeCommandStore safeStore, SafeCommand safeCommand, Truncate truncate, boolean notifyListeners)
+    {
+        Command command = safeCommand.current();
+        Invariants.checkState(!command.is(Truncated));
+
+        //   1) a command has been applied; or
+        //   2) has been coordinated but *will not* be applied (we just haven't witnessed the invalidation yet); or
+        //   3) a command is durably decided and this shard only hosts its home data, so no explicit truncation is necessary to remove it
+        // TODO (desired): consider if there are better invariants we can impose for undecided transactions, to verify they aren't later committed (should be detected already, but more is better)
+        Invariants.checkState(command.hasBeen(Applied) || !command.hasBeen(PreCommitted) || command.partialTxn().keys().isEmpty());
```

Comment:
```suggestion
        Invariants.checkState(command.hasBeen(Applied) || !command.hasBeen(PreCommitted) || command.partialTxn().keys().isEmpty(), "Command %s could not be truncated", command);
```

Reply:
Should this delegate to `shouldTruncate`?  There are subtle differences, so feels like we are drifting (we can set something `Truncated` when `shouldTrucate(cmd) == No`)...

Reply:
It is possible to truncate commands for which shouldTruncate(command) would be No, for instance when we discover that the command cannot have reached consensus globally. This is what `!command.hasBeen(PreCommitted)` covers. However, if you like we can perhaps create a shared method that checks the common invariants, so it's clear what is unique in each case.